### PR TITLE
[train][torchft][data] Replacement worker gets replaced worker’s data split

### PR DESCRIPTION
**Pull request by TimothySeah**
# Summary

This is a continuation of the ray train + torchft integration (https://docs.google.com/document/d/1phjJanIPv5JSD-_B-iPp4Afa4cRV_sEBLeW6q_Z26Xc/edit?tab=t.0#heading=h.mm9jf5hqeinp) to also work seamlessly with Ray Data. When a training worker dies, the train controller spins up a replacement worker that loads the model from another worker (already works with train + torchft integration) and loads the dataloader state from the train controller (this PR) - all without restarting the worker group or saving/loading checkpoints.

It works on three levels:
1) Pass "live" dataset shard to replacement worker. `DatasetsCallback` holds a reference to each worker's `RayDatasetShardProvider`, so it can simply use `before_init_train_context_on_replica_group` to give the replacement worker that shard.
2) Rejoin epoch. Ray Data's `SplitCoordinator` has an "epoch barrier" such that we only begin dataset execution for an epoch after every worker has joined the barrier. `before_init_train_context_on_replica_group` calls the new `set_rejoined_epoch` method to avoid starting a new epoch barrier on joining.
3) 2 phase commit to avoid data loss. Right now, as soon as the `SplitCoordinator` sends a block to the worker, it marks that block as "done." However, you can get into a situation where a worker prefetches more blocks than it has processed and dies, so those blocks are lost. We get around this problem with 2 phase commit, which I explain below.

2 phase commit can be summarized by this diagram:

  queue ──get──► reserved(consumer_id, offset=0) ──ack(k)──► committed up to row k
                         │
                         └── disconnect/abort ──► returned to queue with offset=k

It is implemented at the following levels:
1) 2pc methods on `SplitCoordinator`: `reserve` moves blocks from `_next_bundle` to the `_reserved` queue, `ack` pops/updates blocks and their offsets (we care about the number of rows that have been read for each block) from the `_reserved` queue, and `abort` moves the dead consumer's (block, offset)'s from `_reserved` to `_requeue` (it also slices partially consumed blocks and `ray.put`'s the sliced block)
2) `RefBundle` state management. `iter_batches` creates a `RefBundle` iterator where each `RefBundle` is populated by `gen_blocks` which in turn calls gets the blocks from `SplitCoordinator.get`. Now, `SplitCoordinator.get` tries to get the `RefBundle` from `requeue` (basically, the previously aborted bundles) before getting it from `next_bundle`/`get_next` like it does today. Then, after getting the bundle, `SplitCoordinator` calls `reserve` on it (as opposed to discarding it like it does today). Finally, when the train controller detects that a worker has died, it calls the `abort` method above, which populates `requeue` in the first place.
3) Batch provenance on `StreamSplitDataIterator`. The `StreamSplitDataIterator` formats the `RefBundle`s from the iterator above into batches. The `Batch` dataclass now contains provenance (block_id, rows taken) information. The worker forms batches with `_BatchIteratorWithProvenance` (basically the same as `_BatchIterator` except it also tracks provenance - this occurs within `_pipeline`) and calls the `ack` method (we call `_ack_callback` after getting the batch) above with this provenance information right before yielding the batch. This achieves at most once semantics - if the worker yields the batch and dies before using it, the batch will not be replayed, which is preferable to nccl hanging at the end of training due to an uneven number of batches. We can consider supporting exactly once semantics with dataset rewinding in the future.

We raise an error if the user uses both `local_shuffle_buffer_size` and `enable_2pc_batch_consumption_tracking`. This is fine because we are advising users to use `map_batches` for shuffling instead - CC @xinyuangui2 

# Testing

TODO: Running a benchmark in RayTurbo.

### CODE REVIEW

File: python/ray/train/v2/_internal/callbacks/datasets.py [hunk 1]
```diff
@@ -145,29 +159,112 @@ def before_init_train_context(
             RayDatasetShardProvider(ds_iterators=ds_iterators_per_rank[rank])
             for rank in range(world_size)
         ]
+        self._shard_providers = shard_providers_per_rank
+        self._shard_provider_active = [True] * world_size
         return {"dataset_shard_provider": shard_providers_per_rank}
 
-    def after_worker_group_start(self, worker_group: WorkerGroup):
-        # Propagate DataContext
+    def after_worker_group_shutdown(
```
Comment (TimothySeah) [inline python/ray/train/v2/_internal/callbacks/datasets.py:166]: Assert we are doing fixed number of workers for training since the case where we have (2, 4) workers and we continue training with 3 workers will break this

File: python/ray/train/v2/_internal/callbacks/datasets.py [hunk 2]
```diff
@@ -145,29 +159,112 @@ def before_init_train_context(
...
+            self._shard_provider_active[rank] = True
+            provider = self._shard_providers[rank]
+            # Mark each StreamSplitDataIterator so the replacement worker
+            # rejoins the in-progress epoch instead of arriving at the
+            # start_epoch barrier the predecessor already passed (which would
+            # deadlock the SplitCoordinator). Flag is one-shot: the iterator
+            # auto-clears it after the first iter_batches() call.
+            for ds_iter in provider._dataset_iterators.values():
+                if isinstance(ds_iter, StreamSplitDataIterator):
+                    ds_iter._is_replacement = True
```
Comment (TimothySeah) [inline python/ray/train/v2/_internal/callbacks/datasets.py]: don't access private variable

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 3]
```diff
@@ -47,44 +51,101 @@ def create(
...
     ) -> Tuple[Iterator[RefBundle], Optional[DatasetStats], bool, None]:
+        is_replacement = self._is_replacement
+        self._is_replacement = False
+
         def gen_blocks() -> Iterator[RefBundle]:
-            logger.debug(f"Split {self._output_split_idx}: requesting new epoch.")
-            cur_epoch = ray.get(
-                self._coord_actor.start_epoch.remote(self._output_split_idx)
-            )
+            if is_replacement:
```
Comment (TimothySeah) [inline python/ray/data/_internal/iterator/stream_split_iterator.py]: https://github.com/ray-project/ray/pull/62949 has logic to shut down executor on splitcoordinator when all workers are done. There's also some try finally logic that might accidentally end the epoch even though the new logic is trying to rejoin and finish the epoch.

File: python/ray/data/_internal/block_batching/interfaces.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import abc
-from dataclasses import dataclass
```
Comment (TimothySeah) [inline python/ray/data/_internal/block_batching/interfaces.py:2]: Note: try this with rayturbo too

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 5]
```diff
@@ -62,25 +63,84 @@ def __init__(
...
+                return
+
             # Initial get with 0 prefetched bytes.
             future: ObjectRef[Optional[RefBundle]] = self._coord_actor.get.remote(
                 cur_epoch, self._output_split_idx, 0
             )
-            last_log_time = 0.0
             while True:
                 block_ref_and_md: Optional[RefBundle] = ray.get(future)
                 if not block_ref_and_md:
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/data/_internal/iterator/stream_split_iterator.py]: oudated.

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 6]
```diff
@@ -242,6 +302,26 @@ def start_epoch(self, split_idx: int) -> str:
         epoch_id = self._barrier(split_idx)
         return epoch_id
 
+    def rejoin_current_epoch(self, split_idx: int) -> int:
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/data/_internal/iterator/stream_split_iterator.py]: done.

File: python/ray/data/_internal/iterator/stream_split_iterator.py [hunk 7]
```diff
@@ -62,25 +63,84 @@ def __init__(
...
-            )
+            if is_replacement:
+                logger.debug(
+                    f"Split {self._output_split_idx}: rejoining in-progress epoch."
+                )
+                cur_epoch = ray.get(
+                    self._coord_actor.rejoin_current_epoch.remote(
+                        self._output_split_idx
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (TimothySeah) [reply python/ray/data/_internal/iterator/stream_split_iterator.py]: done.

File: python/ray/train/v2/_internal/callbacks/datasets.py (hunk 1 above)
Comment (TimothySeah) [reply python/ray/train/v2/_internal/callbacks/datasets.py:166]: Added comment. It's up to caller (in this case, `worker_group.py`) to know not to call `replace_replica_group`.

File: python/ray/train/v2/_internal/callbacks/datasets.py (hunk 2 above)
Comment (TimothySeah) [reply python/ray/train/v2/_internal/callbacks/datasets.py]: Done. `rejoined_epoch` is clearer too.
